core/trino-main/src/main/java/io/trino/operator/output/BlockBuilderPositionsAppender.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.output;
+
+import io.trino.spi.block.Block;
+import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.block.BlockBuilderStatus;
+import io.trino.spi.block.DictionaryBlock;
+import io.trino.spi.type.Type;
+import io.trino.type.BlockTypeOperators.BlockPositionEqual;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import org.openjdk.jol.info.ClassLayout;
+
+import static java.util.Objects.requireNonNull;
+
+public abstract class BlockBuilderPositionsAppender
+        implements BlockTypeAwarePositionsAppender
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(BlockBuilderPositionsAppender.class).instanceSize();
+
+    protected final Type type;
+    protected final BlockPositionEqual equalOperator;
+    private final BlockBuilder blockBuilder;
+
+    BlockBuilderPositionsAppender(Type type, BlockPositionEqual equalOperator, BlockBuilderStatus blockBuilderStatus, int expectedPositions)
+    {
+        this.type = requireNonNull(type, "type is null");
+        this.blockBuilder = type.createBlockBuilder(blockBuilderStatus, expectedPositions);
+        this.equalOperator = requireNonNull(equalOperator, "equalOperator is null");
+    }
+
+    public static class TypeBlockBuilderPositionsAppender
+            extends BlockBuilderPositionsAppender
+    {
+        TypeBlockBuilderPositionsAppender(Type type, BlockPositionEqual equalOperator, BlockBuilderStatus blockBuilderStatus, int expectedPositions)
+        {
+            super(type, equalOperator, blockBuilderStatus, expectedPositions);
+        }
+
+        @Override
+        protected void appendTo(IntArrayList positions, Block source, BlockBuilder blockBuilder)
```

Comment:
It's really confusing that `BlockBuilderPositionsAppender` has internal `BlockBuilder`, but also accepts `BlockBuilder`. 

Let's not mix interfaces If `TypeBlockBuilderPositionsAppender` is the only special thingy that requires `BlockBuilder`. IMO it should continue to be `PositionsAppender` until unwrapping of RLE/Dictionary is added.

We don't really need `BlockBuilder` argument anyway when we unwrap RLE/dictionary before calling `TypeBlockBuilderPositionsAppender`

Reply:
it was done to limit the changes to the existing dedicated `PositionAppender`s but I agree, it's better to just access the field. I will refactor this.